src/autoscaler/metricsforwarder/server/custom_metrics_handlers.go

```diff
@@ -0,0 +1,208 @@
+		return true
+	}
+	mh.logger.Debug("failed-to-authorize-credentials")
+	return false
+}
+
+func (mh *CustomMetricsHandler) validateCustomMetricTypes(appGUID string, metricsConsumer *models.MetricsConsumer) (bool, error) {
+	standardMetricsTypes := make(map[string]struct{})
+	standardMetricsTypes[models.MetricNameMemoryUsed] = struct{}{}
+	standardMetricsTypes[models.MetricNameMemoryUtil] = struct{}{}
+	standardMetricsTypes[models.MetricNameThroughput] = struct{}{}
+	standardMetricsTypes[models.MetricNameResponseTime] = struct{}{}
+	standardMetricsTypes[models.MetricNameCPUUtil] = struct{}{}
+
+	allowedMetricTypeSet := make(map[string]struct{})
+	res, found := mh.allowedMetricCache.Get(ALLOWED_METRIC_TYPES)
+	if found {
+		// AllowedMetrics found in cache
+		allowedMetricTypeSet = res.(map[string]struct{})
+	} else {
+		//  AllowedMetrics not found in cache, find AllowedMetrics from Database
+		scalingPolicy, err := mh.policyDB.GetAppPolicy(appGUID)
+		if err != nil {
+			mh.logger.Error("error-getting-policy", err, lager.Data{"appId": appGUID})
+			return false, errors.New("not able to get policy details")
+		}
+		if err == nil && scalingPolicy == nil {
+			mh.logger.Debug("no-policy-found", lager.Data{"appId": appGUID})
+			return false, errors.New("no policy found")
+		}
+		for _, metrictype := range scalingPolicy.ScalingRules {
+			allowedMetricTypeSet[metrictype.MetricType] = struct{}{}
+		}
+		//update the cache
+		mh.allowedMetricCache.Set(ALLOWED_METRIC_TYPES, allowedMetricTypeSet, mh.cacheTTL)
+	}
+	mh.logger.Debug("allowed-metrics-types", lager.Data{"metrics": allowedMetricTypeSet})
+	for _, metric := range metricsConsumer.CustomMetrics {
+		// check if the custom metric name matches with any standard metrics name
+		_, stdMetricsExists := standardMetricsTypes[metric.Name]
+		if stdMetricsExists { //it should fail
+			mh.logger.Info("custom-metric-name-matches-with-standard-metrics-name", lager.Data{"metric": metric.Name})
+			return false, errors.New("CustomMetric name matches with standard metrics name")
+		}
+		// check if any of the custom metrics not defined during policy binding
+		_, ok := allowedMetricTypeSet[metric.Name]
+		if !ok { // If any of the custom metrics is not defined during policy binding, it should fail
+			mh.logger.Info("unmatched-custom-metric-type", lager.Data{"metric": metric.Name})
+			return false, errors.New("CustomMetric name does not match with metric defined in policy")
+		}
+	}
+	return true, nil
+}
+
+func (mh *CustomMetricsHandler) getMetrics(appID string, metricsConsumer *models.MetricsConsumer) []*models.CustomMetric {
+	var metrics []*models.CustomMetric
+	for _, metric := range metricsConsumer.CustomMetrics {
+		metrics = append(metrics, &models.CustomMetric{
+			AppGUID:       appID,
+			InstanceIndex: metricsConsumer.InstanceIndex,
+			Name:          metric.Name,
+			Value:         metric.Value,
+			Unit:          metric.Unit,
+		})
+	}
+	return metrics
```

Comment:
why make a copy of CustomMetrics here?  could we just use  range metricsConsumer.CustomMetrics directly when EmitMetric?

Reply:
Wanted to maintain the modularity here.